trecs/models/recommender.py

```diff
@@ -419,32 +477,21 @@ def recommend(
             if item_indices.shape[1] < num_new_items:
                 self.log("Insufficient number of items left!")
 
-        new_items = np.array([]).reshape((self.num_users, 0)).astype(int)
-        if num_new_items:
-            # no guarantees that randomly interleaved items do not overlap
-            # with recommended items
-            # also possible that item indices are repeated
-            col = self.random_state.integers(
-                item_indices.shape[1], size=(self.num_users, num_new_items)
-            )
-            row = np.repeat(self.users.user_vector, num_new_items).reshape((self.num_users, -1))
-            new_items = item_indices[row, col]
-
-        items = np.zeros((self.num_users, self.num_items_per_iter), dtype=int)
-        # generate indices for recommended and randomly interleaved columns
-        col_idxs = np.zeros(self.num_items_per_iter, dtype=bool)
-        rand_col_idxs = self.random_state.choice(
-            self.num_items_per_iter, size=num_new_items, replace=False
-        )
-        col_idxs[rand_col_idxs] = True
-        items[:, ~col_idxs] = recommended  # preserving relative order of recommended items
-        items[:, col_idxs] = new_items
+        interleaved_items = self.choose_interleaved_items(num_new_items, item_indices)
+
+        if num_new_items > 0:
+            items = self.random_state.random((self.num_users, self.num_items_per_iter))
+            interleave_mask = np.zeros(items.shape).astype(bool)
+            rand_col_idxs = items.argpartition(-num_new_items)[:, -num_new_items:]
+            np.put_along_axis(interleave_mask, rand_col_idxs, True, axis=1)
+            items[interleave_mask] = interleaved_items.flatten()
+            items[~interleave_mask] = recommended.flatten()
```

Comment:
Related to above comment. Couldn't we pop a set in here? Something like `items[~iterleave_mask] = np.setxor1d(interleaved_items.flatten(), recommended.flatten())`. That syntax probably doesn't work, but hopefully it illustrates the point.

Reply:
One potential issue (maybe @sunnymatt can let us know if this is correct) is that, when excluding items with `set`, you might end up with something that is not a matrix. This would happen in the case of an unequal number of repetitions for each user (e.g., `user1` has 2 repeated items, `user2` has 1 repeated item; when you remove the repetitions, you are left with a different number of items for `user1` and `user2`, and numpy will not like it).

Reply:
I see. If that is true, I can understand this approach, but I'd like to then file an issue to figure out a solution to the problem.

Reply:
Yeah, I've filed an issue and have a decent idea of how to avoid this issue! Stay tuned!